### ISSUE TITLE
rustdoc intra-doc links emits spurious warnings

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/rust#83677]
```rust
//! <code>T::[from][`From::from`](x)</code>
```

When generating docs, this emits the warning:
```
warning: unresolved link to `x`
 --> src/lib.rs:1:35
  |
1 | //! <code>T::[from][`From::from`](x)</code>
  |                                   ^ no item named `x` in scope
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

warning: 1 warning emitted
```

I get no warning for this:
```rust
//! <code>T::[from][`From::from`](x)</code>
//!
//! [`From::from`]: `From::from`
```
and no warning for this:
```rust
//! <code>T::[from][`From::from`]\(x)</code>
```

### Meta
`rustc +nightly --version`:
```
rustc 1.53.0-nightly (4a20eb6a9 2021-03-28)
```

**Comment by jyn514** [comment]
@tspiteri the warning is correct. Here's the HTML generated for each:
```html
<code>T::<a href="../core/convert/trait.From.html#tymethod.from" title="From::from">from</a>(x)</code>
<code>T::<a href="../core/convert/trait.From.html#tymethod.from" title="From::from">from</a>(x)</code>
<code>T::<a href="../core/convert/trait.From.html#tymethod.from">from</a>(x)</code>
```
The issue is that `[a][b](c)` is ambiguous markdown syntax: it could either be
1. a link to `b` titled `a`, which you seemed to intend, or
2. the literal text `[a]`, with a link to `c` titled `b`.

Rustdoc is using the second meaning here, and when `c` fails to resolve it falls back to the first meaning. I'd recommend using `\(x)` or ` (x)`, which disambiguates that you mean 1. Compare that to how it behaves on normal links (ignoring intra-doc links for now):

```rust
//! [a][http://example.com](http://example.com)

/// [a][http://example.com](http://example.com)
///
/// [http://example.com]: http://example.com
pub fn bar() {}

/// [a][http://example.com]\(http://example.com)
pub fn foo() {}
```

![image](https://user-images.githubusercontent.com/23638587/113009909-64370e00-9146-11eb-91fe-40812e8201a4.png)

Note that the crate-level docs use meaning 2.

**Comment by Nemo157** [comment]
> Rustdoc is using the second meaning here, and when `c` fails to resolve it falls back to the first meaning.

That seems incorrect based on my reading of the CommonMark Spec, (though, since it defines the parser largely through examples and doesn't have an example exactly like this, I can't be sure). The closest examples are [examples 565-567](https://spec.commonmark.org/0.29/#example-565). From these my expectations for how `[a][b](c)` is parsed is:

1. if `b` is a link reference: `full-reference-link("a", "b"), text("(c)")`
2. otherwise; `text("[a]") inline-link("b", "c")`

A complicating factor here is that `rustdoc` introduces "implicit reference link definitions" for intra-rustdoc-links; given that those result in option 1 being picked, they should not attempt to be resolving a link to `c` and emitting a warning, since it has been chosen that `(c)` is just text.

**Comment by Nemo157** [comment]
```rust
//! [a][b](c)
//! [d][main](e)
//! [f][g](main)
pub fn main() {}
```

```
warning: unresolved link to `b`
1 | //! [a][b](c)
  |         ^ no item named `b` in scope
warning: unresolved link to `g`
3 | //! [f][g](main)
  |         ^ no item named `g` in scope
warning: unresolved link to `c`
1 | //! [a][b](c)
  |            ^ no item named `c` in scope
warning: unresolved link to `e`
2 | //! [d][main](e)
  |               ^ no item named `e` in scope
```

Of these warnings I think the final one should not be emitted. The first line gets two warnings emitted because it attempts to first resolve the full-reference-link `[a][b]`, fails, and switches to instead resolving the inline-link `[b](c)`, and fails again. The third line gets 1 warning because it attempts to first resolve the full-reference-link `[f][g]`, fails, and switches to instead resolving the inline-link `[g](main)` which succeeds. The second line should have no warnings because it first attempts to resolve the full-reference-link `[d][main]`, succeeds and then treats the `(e)` as text.

**Comment by jyn514** [comment]
> In the following case [bar][baz] is parsed as a reference, [foo] as normal text:
> `[foo][bar][baz]`

Hmm, that makes me think meaning 2 should have precedence over meaning 1.

**Comment by jyn514** [comment]
I don't think there's much rustdoc can do here - "markdown is poorly specified" isn't something we can fix.
